wallet-new/test/unit/Wallet/Abstract/Cardano.hs

```diff
@@ -0,0 +1,272 @@
+{-# LANGUAGE UndecidableInstances #-}
+
+module Wallet.Abstract.Cardano (
+    -- * Cardano interpreter for the inductive wallet
+    InductiveT(..)
+  , interpretT
+    -- * Equivalence check
+  , EquivalenceViolation(..)
+  , EquivalenceViolationEvidence(..)
+  , equivalentT
+  ) where
+
+import           Universum
+import qualified Data.Text.Buildable
+import           Formatting (bprint, build, (%))
+import           Pos.Txp (Utxo, formatUtxo)
+import           Pos.Core (HasConfiguration, Coin, unsafeIntegerToCoin)
+import           Pos.Crypto (EncryptedSecretKey)
+import qualified Cardano.Wallet.Kernel as Kernel
+import           Cardano.Wallet.Kernel.Types
+
+import           Util.Validated
+import           UTxO.Context (Addr)
+import           UTxO.DSL (Hash)
+import qualified UTxO.DSL as DSL
+import           UTxO.Interpreter
+import           UTxO.Translate
+import           Wallet.Abstract
+
+{-------------------------------------------------------------------------------
+  Interpreter for the wallet using the translated Cardano types
+-------------------------------------------------------------------------------}
+
+-- | Callbacks used in 'interpretT'
+--
+-- We do not run the callback in the 'IntT' monad so that we maintain
+-- control over the interpretation context.
+data InductiveT h m = InductiveT {
+      -- | Initialize the wallet
+      --
+      -- The callback is given the translated UTxO of the bootstrap
+      -- transaction (we cannot give it the translated transaction because
+      -- we cannot translate the bootstrap transaction).
+      walletBootT :: InductiveCtxt h -> Utxo -> m Kernel.WalletId
+
+      -- | Apply a block
+    , walletApplyBlockT :: InductiveCtxt h -> Kernel.WalletId -> RawResolvedBlock -> m ()
+
+      -- | Insert new pending transaction
+    , walletNewPendingT :: InductiveCtxt h -> Kernel.WalletId -> RawResolvedTx -> m ()
+    }
+
+-- | The context in which a function of 'InductiveT' gets called
+data InductiveCtxt h = InductiveCtxt {
+      -- | The 'Inductive' value that led to this point
+      inductiveCtxtInd :: Inductive h Addr
+
+      -- | The 'IntCtxt' suitable for translation derived values
+      -- (such as UTxOs)
+    , inductiveCtxtInt :: IntCtxt h
+
+      -- | The pure wallet value at this point
+    , inductiveCtxtWallet :: Wallet h Addr
+    }
+
+-- | Interpreter for inductive wallets using the translated Cardano types
+interpretT :: forall h e m. (Monad m, Hash h Addr)
+           => (DSL.Transaction h Addr -> Wallet h Addr)
+           -> InductiveT h (TranslateT e m)
+           -> Inductive h Addr
+           -> TranslateT (Either IntException e) m (Wallet h Addr, IntCtxt h)
+interpretT mkWallet InductiveT{..} ind'' =
+    -- This is ugly, but we only discover the bootstrap transaction once we
+    -- descend down the 'Inductive' wallet. We will 'put' the right context
+    -- before the first call to 'int'.
+    runIntT (error "initialized later") (fst <$> go ind'')
```

Comment:
Can we embed the comment into the error message? eg `error "You called 'int' before the context was initialized."`

Reply:
Changed it to `interpretT: the impossible happened`. Explaining the full setup isn't very useful here, I think, but at least this is a lot more greppable if/when it would ever arise.